darkdetect/_mac_detect.py

```diff
@@ -71,54 +75,67 @@ def theme():
 
     msg(pool, n('release'))
 
-    if out is not None:
-        return out.decode('utf-8')
-    else:
-        return 'Light'
-
-def isDark():
-    return theme() == 'Dark'
-
-def isLight():
-    return theme() == 'Light'
+    return "Light" if out is None else out.decode('utf-8')
 
 
-def _listen_child():
+class MacListener(BaseListener):
     """
-    Run by a child process, install an observer and print theme on change
+    A listener class for macOS
     """
-    import signal
-    signal.signal(signal.SIGINT, signal.SIG_IGN)
-
-    OBSERVED_KEY = "AppleInterfaceStyle"
-
-    class Observer(NSObject):
-        def observeValueForKeyPath_ofObject_change_context_(
-            self, path, object, changeDescription, context
-        ):
-            result = changeDescription[NSKeyValueChangeNewKey]
-            try:
-                print(f"{'Light' if result is None else result}", flush=True)
-            except IOError:
-                os._exit(1)
-
-    observer = Observer.new()  # Keep a reference alive after installing
-    defaults = NSUserDefaults.standardUserDefaults()
-    defaults.addObserver_forKeyPath_options_context_(
-        observer, OBSERVED_KEY, NSKeyValueObservingOptionNew, 0
-    )
-
-    AppHelper.runConsoleEventLoop()
-
-
-def listener(callback: Callable[[str], None]) -> None:
-    if not _can_listen:
-        raise NotImplementedError()
-    with subprocess.Popen(
-        (sys.executable, "-c", "import _mac_detect as m; m._listen_child()"),
-        stdout=subprocess.PIPE,
-        universal_newlines=True,
-        cwd=Path(__file__).parent,
-    ) as p:
-        for line in p.stdout:
-            callback(line.strip())
+
+    def __init__(self, callback: Callable[[str], None]):
+        self._proc: Optional[subprocess.Popen] = None
+        super().__init__(callback)
+
+    # Overrides
+
+    def _listen(self):
+        if not _can_listen:
+            raise NotImplementedError("Optional dependencies not found; fix this with: pip install darkdetect[macos-listener]")
+        self._proc = subprocess.Popen(
+                (sys.executable, "-c", "import darkdetect as d; d.MacListener._listen_child()"),
```

Comment:
Once again, this can cause different results than expected if the installed copy differs from the local copy. I would make this always local by importing directly from `_mac_detect` as we had agreed in the previous PR.

Reply:
I did this since `_mac_detect.py` now imports `from .`; and you will get an error if importing like this:
```python
>>> from ._base_listener import BaseListerner
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: attempted relative import with no known parent package
```

This should be safe though, items in the current working directory are prioritized. You can check this via `pip install darkdetect` then `cd`ing into a `darkdetect` clone, editing `__init__.py` to `print('foo')`, then `python -c 'import darkdetect'`.

Reply:
I know that the current working directory should always be prioritized in `sys.path`, but sometimes `sys.path` can be altered, e.g. by `py2app`. I do not intend to use the listener in frozen apps, but some users might and I am not 100% sure this code would work in such case. I guess I will have to add this to the list of tests I need to run on this PR.

Reply:
If this is an issue, I can go back to having the subprocess command contain `sys.path.insert(0, Path(__file__).parents[1])` instead of setting the working directory.

Reply:
Hmm, so I have some bad news. I've spent quite a bit of time think / tinkering on this but I'm not sure how to get this (or what currently exists) to work with all types of builds.

The big issue is the frozen apps which lack a python interpreter (for example, pyinstaller builds may not have a public python interpreter, because the interpreter along with many related files have been all packaged into a single executable).

The existing implementation assumes `sys.executable` is a python interpreter which exposes `-c`. `pyinstaller`, for example, does not; it does expose `-i` but that is not the same.  

To get around this I did try forking, but recent versions of macOS limit forking in use with objc code, so that won't work.

`multiprocessing` has frozen support and can do `spawn` in lieu of `fork`, but that introduces its own problems.

I will give another shot at getting the code to run in a non-main thread.

Reply:
With that said, that might be a different PR; since this PR is less about changing how the listener listens and more just adapting the API.

Reply:
I can add a note to the README about this though, for now at least.

Reply:
I completely agree that the issues with `pyinstaller` and `py2app` do not belong to this PR. I am fine with mentioning in the README that the macOS listener does not currently work in frozen `.app` bundles. 

Side note: `py2app` ships an executable in the bundles, so I guess this would be possible, but I have not investigated this yet.

Reply:
I can open a related issue once this is merged. I've had no luck with observers in different threads, though I also don't know Objective C so reading such code has been tricky for me. I found some posts mentioning this might be possible via the notification API. I tested in the main loop and stuff like this works:
```python
  ...
    nc = Foundation.NSDistributedNotificationCenter.defaultCenter()
    nc.addObserver_selector_name_object_(self, 'mycallback:', 'AppleInterfaceThemeChangedNotification',None)

def mycallback_(self, value): 
    return self._actual_callback(value)
 ```
 There are some functions about adding an observer to a specific RunLoop, which might be what is needed, to have this observer use a thread's run loop? Though maybe the existing method of observation is sufficient.
 
If it is not possible, polling might be the only option though :( I did time it and polling takes almost no resources, it is also maybe worth mentioning that `subprocess.Popen` also does polling to ensure that its child is alive and isn't replaced by another process of the same `pid`; but this solution is still one to be avoided. Perhaps if a PR allows polling, it could extent the `BaseListener` class by modifying `__init__` to take an argument `__init__(self, callback, fallback_to_polling=False)`? That way users would have to be explicit about enabling it and it would change no existing functionality as the default would be `False`. I think that's an acceptable that might be tolerable in the meantime, until a better solution for `pyinstaller`-type programs is found (in my case, `pyinstaller` is one of the types I need this to work with :( ).

In any case, this is a different PR, so I'll mark this resolved for now.